File: frontend/src/domain/finance.ts
/** TypeScript interfaces mirroring the finance domain entities. */

export type AccountType = 'bank' | 'cash' | 'credit_card' | 'savings' | 'investment'
export type TransactionType = 'income' | 'expense' | 'transfer'
export type CategoryType = 'income' | 'expense' | 'transfer'

export interface Account {
  id: string
  name: string
  account_type: AccountType
  currency: string
  is_active: boolean
  created_at: string
}

export interface TransactionCategory {
  id: string
  name: string
  category_type: CategoryType
  parent_id: string | null
  color: string | null
  icon: string | null
  sort_order: number
  is_system: boolean
  children: TransactionCategory[]
}

export interface Transaction {
  id: string
  account_id: string
  category_id: string | null
  payment_method: string | null
  counterparty_ref: string | null
  counterparty_name: string | null
  transfer_pair_id: string | null
  amount_thb: number
  original_amount: number | null
  original_currency: string | null
  transaction_type: TransactionType
  description: string
  transaction_date: string
  tags: string[]
  is_recurring: boolean
  /** OCR metadata — includes transaction_time (HH:MM) when imported from a PDF */
  metadata: Record<string, unknown> | null
  created_at: string
}

export interface TransactionListResponse {
  items: Transaction[]
  total: number
  limit: number
  offset: number
}

export interface TransactionComment {
  id: string
  transaction_id: string
  content: string
  created_at: string
  updated_at: string
}

export interface TransactionGroup {
  id: string
  name: string
  description: string | null
  color: string | null
  created_at: string
}

// Flow tree types for visualization
export interface FlowNode {
  id: string
  label: string
  node_type: 'income_source' | 'account' | 'expense_category' | 'transfer'
  total_thb: number
  color: string | null
  icon: string | null
}

export interface FlowEdge {
  source_id: string
  target_id: string
  amount_thb: number
  label: string | null
}

export interface FlowTree {
  nodes: FlowNode[]
  edges: FlowEdge[]
  total_income_thb: number
  total_expense_thb: number
  net_thb: number
}

// ── Summary types ─────────────────────────────────────────────────────────────

export interface CategoryStat {
  category_id: string
  name: string
  color: string | null
  icon: string | null
  total_thb: number
  count: number
  percentage: number
}

export interface MonthlyPoint {
  month: string
  income_thb: number
  expense_thb: number
  net_thb: number
  count: number
}

export interface AccountStat {
  account_id: string
  name: string
  account_type: AccountType
  currency: string
  balance_thb: number
  period_income_thb: number
  period_expense_thb: number
}

export interface PaymentMethodStat {
  method: string
  total_thb: number
  count: number
  percentage: number
}

export interface Summary {
  date_from: string | null
  date_to: string | null
  total_income_thb: number
  total_expense_thb: number
  net_thb: number
  transaction_count: number
  uncategorized_count: number
  recurring_count: number
  monthly_trend: MonthlyPoint[]
  top_expense_categories: CategoryStat[]
  top_income_categories: CategoryStat[]
  accounts: AccountStat[]
  payment_methods: PaymentMethodStat[]
}

File: backend/src/pbam/interfaces/api/v1/schemas/finance.py
"""Pydantic v2 schemas for finance endpoints."""
from datetime import date, datetime
from decimal import Decimal
from typing import Any
from uuid import UUID

from pydantic import BaseModel, field_validator


# ── Accounts ──────────────────────────────────────────────────────────────────

class AccountCreate(BaseModel):
    name: str
    account_type: str
    currency: str = "THB"
    initial_balance: Decimal = Decimal("0")
    metadata: dict[str, Any] = {}


class AccountUpdate(BaseModel):
    name: str | None = None
    is_active: bool | None = None
    metadata: dict[str, Any] | None = None


class AccountResponse(BaseModel):
    id: UUID
    name: str
    account_type: str
    currency: str
    is_active: bool
    created_at: datetime

    model_config = {"from_attributes": True}


# ── Categories ────────────────────────────────────────────────────────────────

class CategoryCreate(BaseModel):
    name: str
    category_type: str
    parent_id: UUID | None = None
    color: str | None = None
    icon: str | None = None
    sort_order: int = 0


class CategoryUpdate(BaseModel):
    name: str | None = None
    color: str | None = None
    icon: str | None = None
    sort_order: int | None = None
    parent_id: UUID | None = None


class CategoryResponse(BaseModel):
    id: UUID
    name: str
    category_type: str
    parent_id: UUID | None
    color: str | None
    icon: str | None
    sort_order: int
    is_system: bool
    children: list["CategoryResponse"] = []

    model_config = {"from_attributes": True}


# ── Transactions ──────────────────────────────────────────────────────────────

class TransactionCreate(BaseModel):
    account_id: UUID
    amount: Decimal
    currency: str = "THB"
    exchange_rate: Decimal | None = None
    transaction_type: str
    description: str
    transaction_date: date
    category_id: UUID | None = None
    payment_method: str | None = None
    tags: list[str] = []
    is_recurring: bool = False

    @field_validator("transaction_type")
    @classmethod
    def valid_type(cls, v: str) -> str:
        if v not in ("income", "expense", "transfer"):
            raise ValueError("transaction_type must be income, expense, or transfer")
        return v


class TransactionUpdate(BaseModel):
    description: str | None = None
    category_id: UUID | None = None
    transaction_type: str | None = None
    payment_method: str | None = None
    tags: list[str] | None = None
    transaction_date: date | None = None
    counterparty_name: str | None = None

    @field_validator("transaction_type")
    @classmethod
    def valid_type(cls, v: str | None) -> str | None:
        if v is not None and v not in ("income", "expense", "transfer"):
            raise ValueError("transaction_type must be income, expense, or transfer")
        return v


class TransactionResponse(BaseModel):
    id: UUID
    account_id: UUID
    category_id: UUID | None
    payment_method: str | None
    counterparty_ref: str | None
    counterparty_name: str | None
    transfer_pair_id: UUID | None
    amount_thb: Decimal
    original_amount: Decimal | None
    original_currency: str | None
    transaction_type: str
    description: str
    transaction_date: date
    tags: list[str]
    is_recurring: bool
    metadata: dict[str, Any] | None
    created_at: datetime

    model_config = {"from_attributes": True}


class TransactionListResponse(BaseModel):
    items: list[TransactionResponse]
    total: int
    limit: int
    offset: int


# ── Comments ──────────────────────────────────────────────────────────────────

class CommentCreate(BaseModel):
    content: str


class CommentResponse(BaseModel):
    id: UUID
    transaction_id: UUID
    content: str
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# ── Groups ────────────────────────────────────────────────────────────────────

class GroupCreate(BaseModel):
    name: str
    description: str | None = None
    color: str | None = None


class GroupResponse(BaseModel):
    id: UUID
    name: str
    description: str | None
    color: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


# ── Flow tree ─────────────────────────────────────────────────────────────────

class FlowNodeResponse(BaseModel):
    id: str
    label: str
    node_type: str
    total_thb: Decimal
    color: str | None
    icon: str | None


class FlowEdgeResponse(BaseModel):
    source_id: str
    target_id: str
    amount_thb: Decimal
    label: str | None


class FlowTreeResponse(BaseModel):
    nodes: list[FlowNodeResponse]
    edges: list[FlowEdgeResponse]
    total_income_thb: Decimal
    total_expense_thb: Decimal
    net_thb: Decimal
